allennlp/modules/transformer/transformer_module.py

```diff
@@ -1,229 +1,433 @@
-from typing import Optional, Dict, Union, List, Any
+from enum import Enum
 import logging
-import inspect
+import os
+from os import PathLike
+from typing import TYPE_CHECKING, Optional, Dict, Union, List, Any, TypeVar, Type
+import re
+import warnings
 
 import torch
+import torch.distributed as dist
+
+from allennlp.common.util import is_distributed, is_global_primary
+from allennlp.nn.util import StateDictType, read_state_dict, load_state_dict_distributed
+
+if TYPE_CHECKING:
+    from transformers.configuration_utils import PretrainedConfig
 
-from allennlp.common import cached_transformers
 
 logger = logging.getLogger(__name__)
 
 
+_T = TypeVar("_T", bound="TransformerModule")
+
+
+class DistributedLoadingStrategy(Enum):
+    """
+    Strategy options for loading state dictionaries across distributed processes.
+    """
+
+    FREE_FOR_ALL = "FREE_FOR_ALL"
```

Comment:
Hmm I guess I was thinking that since `FREE_FOR_ALL` is faster, it will probably preferred when you have enough memory.

I should benchmark the two approaches. If there's not a significant different for medium sized models then I agree, we should use the memory efficient strategy all the time. I'll get back to you on that.

Reply:
There wasn't much of a difference for `t5-base` on my Mac with 2 workers (3.55 seconds vs 3.87 seconds). So I think it's worth it to always use the memory efficient strategy.